File: core/src/test/java/com/ritense/valtimo/web/rest/SecuritySmokeIntegrationTest.java
/*
 * Copyright 2015-2020 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ritense.valtimo.web.rest;

import com.ritense.valtimo.contract.authentication.UserManagementService;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.TestFactory;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.SecurityConfig;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.request.RequestPostProcessor;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.method.HandlerMethod;
import org.springframework.web.servlet.mvc.condition.NameValueExpression;
import org.springframework.web.servlet.mvc.method.RequestMappingInfo;
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping;
import java.lang.annotation.Annotation;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SpringExtension.class)
@SpringBootTest
@AutoConfigureMockMvc
@Tag("security")
public abstract class SecuritySmokeIntegrationTest {

    @MockBean
    public UserManagementService userManagementService;

    @Autowired
    private RequestMappingHandlerMapping handlerMapping;

    @Autowired
    private MockMvc mockMvc;

    @TestFactory
    public Collection<DynamicTest> shouldTestAllSimpleEndpointsForAuthentication() {
        Set<Map.Entry<RequestMappingInfo, HandlerMethod>> endPointData = handlerMapping.getHandlerMethods().entrySet();
        Set<String> ignoredPathPatterns = getIgnoredPathPatterns();
        List<DynamicTest> dynamicTests = new ArrayList<>(endPointData.size());
        for (Map.Entry<RequestMappingInfo, HandlerMethod> entry : endPointData) {
            dynamicTests.add(getDynamicEndpointTest(entry, ignoredPathPatterns));
        }
        return dynamicTests;
    }

    protected abstract Set<String> getIgnoredPathPatterns();

    private DynamicTest getDynamicEndpointTest(
        Map.Entry<RequestMappingInfo, HandlerMethod> entry,
        Set<String> ignoredPathPatterns
    ) {
        final RequestMappingInfo key = entry.getKey();
        final HandlerMethod handlerMethod = entry.getValue();

        final String methods = key.getMethodsCondition().getMethods().stream().map(Object::toString).collect(Collectors.joining("|"));
        final Set<String> patterns = key.getPatternsCondition().getPatterns();

        String testName = String.format("%s %s", methods, String.join(", ", patterns));

        return DynamicTest.dynamicTest(testName, () -> {
            EndpointSecurityResult.Builder builder = EndpointSecurityResult.builder()
                .methods(key.getMethodsCondition().getMethods())
                .paramConditions(key.getParamsCondition().getExpressions().stream().map(Object::toString).collect(Collectors.toSet()))
                .patterns(patterns)
                .classSecurityAnnotations(getSecurityAnnotations(handlerMethod.getMethod().getDeclaringClass().getAnnotations()))
                .methodSecurityAnnotations(getSecurityAnnotations(handlerMethod.getMethod().getDeclaredAnnotations()))
                .javaMethod(handlerMethod.getMethod().getDeclaringClass().getName() + "." + handlerMethod.getMethod().getName());

            SecurityResult securityResult = testEndpoint(key, true);
            if (securityResult == SecurityResult.INSECURE) {
                builder.openForLocalNetwork(true);
                securityResult = testEndpoint(key, false);
                builder.openForPublicNetwork(securityResult == SecurityResult.INSECURE);
            }

            builder.securityResult(securityResult);

            if (ignoredPathPatterns.containsAll(patterns)) {
                builder.testResult(TestResult.IGNORED);
            } else if (securityResult == SecurityResult.SECURE) {
                builder.testResult(TestResult.PASS);
            } else {
                builder.testResult(TestResult.FAIL);
            }

            EndpointSecurityResult result = builder.build();
            assertThat(result.getTestResult())
                .as(result.toPrettyString())
                .isNotEqualByComparingTo(TestResult.FAIL);
        });
    }

    private Set<String> getSecurityAnnotations(Annotation[] annotations) {
        String packageName = SecurityConfig.class.getPackage().getName();
        return Arrays.stream(annotations)
            .filter(annotation -> annotation.annotationType().getPackage().getName().startsWith(packageName))
            .map(annotation -> annotation.annotationType().getSimpleName() + "(\"" + getAnnotationValue(annotation) + "\")")
            .collect(Collectors.toSet());
    }

    private String getAnnotationValue(Annotation annotation) {
        try {
            return Arrays.stream(annotation.annotationType().getMethods())
                .filter(method -> method.getName().equals("value"))
                .filter(method -> !method.getReturnType().equals(Void.TYPE))
                .findFirst()
                .flatMap(method -> {
                    try {
                        Object valueResult = method.invoke(annotation);
                        if (!method.getReturnType().isArray()) {
                            valueResult = new Object[] {valueResult};
                        }
                        return Optional.of((Object[]) valueResult);
                    } catch (Exception e) {
                        throw new RuntimeException(e);
                    }
                })
                .map(valueObject -> Arrays.stream(valueObject)
                    .filter(Objects::nonNull)
                    .map(Object::toString)
                    .collect(Collectors.joining("\", \"")))
                .orElse("");
        } catch (Exception ex) {
            return "";
        }
    }

    private SecurityResult testEndpoint(RequestMappingInfo mappingInfo, boolean useLocalNetwork) {
        Optional<String> path = findSimplePath(mappingInfo);
        Optional<MultiValueMap<String, String>> paramMap = getSimpleParamMap(mappingInfo);

        if (path.isEmpty()
            || paramMap.isEmpty()) {
            return SecurityResult.UNKNOWN;
        }

        HttpMethod method = mappingInfo.getMethodsCondition().getMethods().stream().findFirst()
            .map(requestMethod -> HttpMethod.valueOf(requestMethod.name()))
            .orElse(HttpMethod.GET);

        try {
            MockHttpServletRequestBuilder request = MockMvcRequestBuilders.request(method, path.get());
            mappingInfo.getConsumesCondition().getConsumableMediaTypes().stream().findFirst()
                .ifPresent(request::contentType);

            request.params(paramMap.get());
            request.accept("*/*");

            request.with(remoteAddr(useLocalNetwork ? "127.0.0.1" : "8.8.8.8"));

            MvcResult mvcResult = mockMvc.perform(request)
                .andReturn();

            int statusCode = mvcResult.getResponse().getStatus();
            if (Arrays.asList(HttpStatus.UNAUTHORIZED.value(), HttpStatus.FORBIDDEN.value()).contains(statusCode)) {
                return SecurityResult.SECURE;
            } else {
                return SecurityResult.INSECURE;
            }
        } catch (Exception ex) {
            return SecurityResult.UNKNOWN;
        }
    }

    private Optional<MultiValueMap<String, String>> getSimpleParamMap(RequestMappingInfo mappingInfo) {
        Set<NameValueExpression<String>> expressions = mappingInfo.getParamsCondition().getExpressions();

        if (expressions.stream()
            .filter(expr -> expr.getValue() != null)
            .anyMatch(expr -> expr.getValue().matches(".*\\{.*\\}.*"))
        ) {
            //I don't know how to resolve these parameters
            return Optional.empty();
        }

        Map<String, List<String>> params = expressions.stream().collect(Collectors.groupingBy(
            NameValueExpression::getName,
            Collectors.mapping(NameValueExpression::getValue, Collectors.toList())));

        return Optional.of(new LinkedMultiValueMap<>(params));
    }

    private static RequestPostProcessor remoteAddr(final String remoteAddr) {
        return request -> {
            request.setRemoteAddr(remoteAddr);
            return request;
        };
    }

    private Optional<String> findSimplePath(RequestMappingInfo key) {
        Set<String> patterns = key.getPatternsCondition().getPatterns();
        return patterns.stream()
            .filter(s -> !s.matches(".*\\{.*:.*\\}.*"))
            .findFirst()
            .map(s -> s.replaceAll("\\{.*\\}", "1337"));
    }

    enum SecurityResult {
        SECURE,
        INSECURE,
        UNKNOWN
    }

    enum TestResult {
        PASS,
        FAIL,
        IGNORED
    }

}

File: test-utils-common/src/main/kotlin/com/ritense/valtimo/web/rest/SecuritySmokeIntegrationTest.kt
/*
 * Copyright 2015-2023 Ritense BV, the Netherlands.
 *
 * Licensed under EUPL, Version 1.2 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ritense.valtimo.web.rest

import com.ritense.valtimo.contract.authentication.UserManagementService
import com.ritense.valtimo.contract.mail.MailSender
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.TestFactory
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.HttpMethod
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.request.RequestPostProcessor
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.util.LinkedMultiValueMap
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.mvc.method.RequestMappingInfo
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping

/**
 * This abstract test class generates security test for each endpoint in the Spring request handler mapping.
 * Each generated test will check if 'some' security has been applied to the endpoint by attempting to perform a request to it.
 * IMPORTANT: The result of this test should be considered indicative. It does not check if the proper authentication or authorization rules have been applied.
 *
 * When performing the requests, required parameters will be filled with a dummy value where possible.
 *
 * @param basePackageName Tests will only be generated for endpoints of which the handlers are defined within the given package. When null, no filter is applied.
 * @param ignoredPathPatterns These request patterns will be ignored. This could be used when an endpoint is insecure on purpose or when the specific requests is too complex to test properly. Patterns typically look like this: 'GET /xyz'
 */
@ExtendWith(SpringExtension::class)
@SpringBootTest
@AutoConfigureMockMvc
@Tag("security")
abstract class SecuritySmokeIntegrationTest(
    private val basePackageName: String? = null,
    private val ignoredPathPatterns: Set<String> = setOf()
) {
    @MockBean
    private lateinit var userManagementService: UserManagementService

    @MockBean
    private lateinit var mailSender: MailSender

    @Qualifier("requestMappingHandlerMapping")
    @Autowired
    private lateinit var handlerMapping: RequestMappingHandlerMapping

    @Autowired
    private lateinit var mockMvc: MockMvc

    @TestFactory
    fun shouldTestAllSimpleEndpointsForAuthentication(): Collection<DynamicTest> {
        val endPointData: Set<Map.Entry<RequestMappingInfo, HandlerMethod>> = handlerMapping.handlerMethods.entries

        return endPointData.filter {
            basePackageName == null || it.value.beanType.packageName.startsWith(basePackageName)
        }.mapNotNull {
            createEndpointTest(it.key, it.value)
        }
    }

    private fun createEndpointTest(
        mappingInfo: RequestMappingInfo, handlerMethod: HandlerMethod
    ): DynamicTest? {

        val testNames = mappingInfo.methodsCondition.methods
            .ifEmpty { setOf(null) }
            .map { it?.toString() ?: "" }
            .sorted()
            .flatMap { method ->
                getPathPatternsCondition(mappingInfo)
                    .sorted()
                    .map { path ->
                        "$method $path"
                    }
            }

        return if (testNames.isEmpty()) {
            null
        } else if (testNames.any { testName -> ignoredPathPatterns.contains(testName) }) {
            DynamicTest.dynamicTest("Ignored: ${testNames.first()}") {}
        } else {
            DynamicTest.dynamicTest(testNames.first()) {
                testEndpoint(mappingInfo, handlerMethod)
            }
        }
    }

    private fun testEndpoint(mappingInfo: RequestMappingInfo, handlerMethod: HandlerMethod) {
        assertDoesNotThrow {
            val path = findSimplePath(mappingInfo)
            val paramMap = getSimpleParamMap(mappingInfo, handlerMethod)

            val method = mappingInfo.methodsCondition.methods.firstOrNull()
                ?.let { HttpMethod.valueOf(it.name) }
                ?: HttpMethod.GET

            val request = MockMvcRequestBuilders.request(method, path)
            mappingInfo.consumesCondition.consumableMediaTypes.firstOrNull()
                ?.run { request.contentType(this) }

            request.params(LinkedMultiValueMap(paramMap))
            request.accept("*/*")
            request.with(remoteAddr())
            val mvcResult = mockMvc.perform(request)
                .andDo(print())
                .andReturn()
            val statusCode = mvcResult.response.status

            assert(statusCode == 401 || statusCode == 403) { "Expected status 401 or 403, was $statusCode." }
        }
    }

    private fun getSimpleParamMap(
        mappingInfo: RequestMappingInfo,
        handlerMethod: HandlerMethod
    ): Map<String, List<String>> {
        val expressions = mappingInfo.paramsCondition.expressions
            .map {
                val value = if (it.value != null && it.value!!.matches(".*\\{.*}.*".toRegex())) {
                    null //We can't guess complex expression values
                } else {
                    it.value
                }
                Pair(it.name, value ?: "")
            }

        val methodParams = handlerMethod.methodParameters.mapNotNull {
            val requestParam = it.getParameterAnnotation(RequestParam::class.java)
            return@mapNotNull requestParam?.let {
                if (!requestParam.required) {
                    null
                } else {
                    val name = if (requestParam.name != "") requestParam.name else requestParam.value
                    Pair(name, "")
                }
            }
        }

        return (expressions + methodParams)
            .groupBy({ it.first }, { it.second })
            .map { entry -> Pair(entry.key, listOf(entry.value.maxOf { it })) }
            .toMap()
    }

    private fun findSimplePath(key: RequestMappingInfo): String {
        val patterns = getPathPatternsCondition(key)
        return patterns.firstOrNull { s -> !s.matches(".*\\{.*:.*}.*".toRegex()) }
            ?.replace("\\{.*}".toRegex(), "1337")
            ?: throw RuntimeException("No simple path found.")
    }

    private fun getPathPatternsCondition(key: RequestMappingInfo): Set<String> {
        return key.pathPatternsCondition
            ?.patterns
            ?.map { it.patternString }
            ?.toSet()
            ?: setOf()
    }

    companion object {
        private fun remoteAddr(remoteAddr: String = "8.8.8.8"): RequestPostProcessor {
            return RequestPostProcessor { request: MockHttpServletRequest ->
                request.remoteAddr = remoteAddr
                request
            }
        }
    }
}